hail/src/main/resources/include/hail/Region2.h

```diff
@@ -0,0 +1,99 @@
+#ifndef HAIL_REGION2_H
+#define HAIL_REGION2_H 1
+
+#include <memory>
+#include <vector>
+#include <utility>
+#include "hail/Upcalls.h"
+#include "hail/NativeStatus.h"
+
+namespace hail {
+
+class RegionPool {
+  static constexpr ssize_t block_size = 64*1024;
+  static constexpr ssize_t block_threshold = 4096;
+
+  public:
+    class Region {
+      public:
+        class SharedPtr {
+          private:
+            Region * region_;
+            void clear();
+          public:
+            SharedPtr(Region * region) : region_(region) {
+              if (region_ != nullptr) { ++(region_->references_); }
+            }
+            SharedPtr(const SharedPtr &ptr) : SharedPtr(ptr.region_) { }
+            SharedPtr(SharedPtr &&ptr) : region_(nullptr) {
+              region_ = ptr.region_;
+              ptr.region_ = nullptr;
+            }
+            ~SharedPtr() { clear(); }
+
+            void swap(SharedPtr &other) { std::swap(region_, other.region_); }
+            SharedPtr& operator=(SharedPtr other) {
+              swap(other);
+              return *this;
+            }
+            SharedPtr& operator=(std::nullptr_t) noexcept {
+              clear();
+              return *this;
+            }
+            inline Region * get() { return region_; }
+            inline Region & operator*() { return *region_; }
+            inline Region * operator->() { return region_; }
+        };
+
+      private:
+        RegionPool * pool_;
+        int references_ {0};
+        size_t block_offset_;
+        std::unique_ptr<char[]> current_block_;
+        std::vector<std::unique_ptr<char[]>> used_blocks_{};
+        std::vector<std::unique_ptr<char[]>> big_chunks_{};
+        std::vector<SharedPtr> parents_{};
+        char * allocate_new_block(size_t n);
+        char * allocate_big_chunk(size_t size);
+      public:
+        Region(RegionPool * pool);
```

Comment:
`Region` should probably have explicitly deleted copy/move constructors/assignment operators.

Reply:
I kind of wang the Region constructor here to be private as well, so that the RegionPool has total control over Region creation, but I'm calling `std::make_unique` to actually construct the object and so it complains. Any suggestions?

Reply:
I'd say this is probably a case where it's okay to sidestep `make_unique` and use the `unique_ptr` constructor directly:
```c++
auto region{  std::unique_ptr<Region>(new Region(pool)) };
```

Reply:
👌